[Spring MVC (인증)] 곰곰 2,3단계 미션 제출합니다.

jaewoo9797 opened the pull request:
안녕하세요 커찬!

이전 미션 마지막에 질문 주신것에 대한 정리
review_1_extension.md , 인증_인가_정리 <- 파일에 생각 정리했는데 마지막에 생각 결론을 아직 정리못하였습니다.
일단 코드로 하면서 더 정리를 했습니다. 이후 관련 정리하며 생각한 부분 추가하겠습니다.


미션을 진행하며 변경점
- 쿠키 매니저 생성해서 쿠키관련 로직을 캡슐화
- 인터셉터와 파라미터리졸버 구현
- 기존에 Reservation 객체를 데이터베이스에 저장하는 파라미터를 수정

jaewoo9797 committed:
Docs : README 파일에 1차 미션 업데이트

jaewoo9797 committed:
Refactor : MemberDao try-catch 구문으로 변경

커찬님의 의견 수용해서 수정. 리스트로 처리하고 스트림으로 처리하는 리소스를 고려하지 못하였음. 기존에 구현했던 방식도 나쁘지 않다고 생각합니다. 보다 가독성이 있다는 측면에서는!

jaewoo9797 committed:
Docs: 1차 리뷰 추가적인 정리를 위해서 파일 생성과 README 파일 업데이트

jaewoo9797 committed:
Refactor: 기존의 Jwt 토큰을 생성하는 createToken 메소드 명을 generateToken 으로 변경

jaewoo9797 committed:
Refactor: 로그인하는 memberService 의 login 반환 타입 변경

기존의 반환 타입인 String 은 토큰을 반환한다는 것을 바로 알 수 없는 문제점을 반환 타입 객체를 만들어서 해결

jaewoo9797 committed:
Refactor: MemberController 상수 분리

jaewoo9797 committed:
Docs : Jwt에 담아야하는 것이 무엇인지에 대해서 정리

클레임의 대한 개념과 리소스를 식별하기 위해서는 어떤 값을 어디에 담아야하는지 정리했습니다.

jaewoo9797 committed:
Refactor: JwtProvider 의 generateToken 메소드에 클레임에 이름과 이메일을 담는 것을 제거했습니다.

sub에 담아주는 엔티티 식별자 값을 담아주어 충분히 로직을 처리할 수 있습니다.

jaewoo9797 committed:
Refactor: dto 관련 패키지 생성하여 패키징

기존의 class 파일로 관리하던 파일을 record로 변경하였습니다.

jaewoo9797 committed:
Docs: 리뷰에 대한 내용 정리, 도메인에 관해서 고민

생각을 해보면서 코드에 적용해보려고 했는데, 생각했던 방법이 좋아보이지 않음.

jaewoo9797 committed:
Refactor: auth를 관리하는 Service 클래스로 역할 분리

jaewoo9797 committed:
Create: 2단계 HandlerMethodArgumentResolver 생성

jaewoo9797 committed:
Create : HandlerMethodArgumentResolver 로직 세부사항 주석으로 순서도 정리

jaewoo9797 committed:
Test: 제공된 테스트 코드 생성

jaewoo9797 committed:
Refactor: Login Member Argument Resolver 토큰값을 찾아서 생성

일단은 기능이 돌아가도록 만들기, 코드 로직을 수정해야함...

jaewoo9797 committed:
Refactor: Reservation Controller 예약 생성 구현

if 분기 처리를 리팩토링 해야함. 일단 돌아가게 구현함.

jaewoo9797 committed:
Refactor: Reservation Service save 메소드 오버로딩

클라이언트의 요청에 이름이 null 일 경우에 다른 매개변수로 저장할 수 있도록 구현

jaewoo9797 committed:
Refactor: ReservationResponse 인스턴스 필드 final 예약어 추가

jaewoo9797 committed:
Refactor: ReservationDao save 메소드 오버로딩

jaewoo9797 committed:
Refactor: AuthService 기존에 클라이언트의 응답 객체로 리턴하는 것을 재사용하지 않고 메소드 생성

클라이언트의 요청에 대한 응답 객체로 변환하는 메소드를 재사용하기 어려워 메소드 하나 생성

jaewoo9797 committed:
Docs: 미션2를 시작하기전에 정리한 것 커밋

미션 시작 전에 정리를 해두었는데, 미션을 하고나서도 잘 모르겠다. 미션 모두 진행한 후에 리팩토링을 하면서 조금 더 생각을 정리할 필요있음!

jaewoo9797 committed:
Feature: RoleInterceptor  권한을 검사하기 위한 인터셉터 생성

jaewoo9797 committed:
Feature: Configuration 클래스에 인터셉터RoleInterceptor 등록

jaewoo9797 committed:
Feature: 권한이 없을 경우에 401 에러를 반환하기 위해서 ExceptionController 에 에러 핸들링 handleRuntimeException 생성

jaewoo9797 committed:
Test: 3단계 미션 테스트코드 업데이트

jaewoo9797 committed:
Refactor: Member Controller Cookie 생성방식 변경

ResponseCookie 를 활용해서 쿠키를 반환해주고 있었는데, logout 처리 로직을 해당 객체로 변경하는 중 의문점을 느끼고 찾아보니 WebFlux 환경에서 ResponseCookie 객체를 사용한다는 것을 알아 변경

jaewoo9797 committed:
Refactor: Resolver, Interceptor 에서 쿠키를 추출해내는 로직의 중복이 생겨서 해당 기능을 담당하는 Extractor 객체를 생성하여 로직 변경

jaewoo9797 committed:
Refactor: WebMvcConfig 루트페이지 상수로 추출

jaewoo9797 committed:
Refactor: Member Controller Cookie 생성 중복 로직 제거

jaewoo9797 committed:
Refactor: Cookie관련 로직을 담당하는 CookieManager 객체

쿠키를 생성해서 response에 담아주고, request에서 token 이름으로 쿠키를 찾도록 수정

jaewoo9797 committed:
이전 커밋에 누락된 클래스 파일 커밋

jaewoo9797 committed:
Refactor: Cookie Manager로 클래스 이름 변경과 함께 사용하는 곳의 변수명 변경, TODO 작성

jaewoo9797 committed:
Refactor: CookieManager 토큰값을 반환하도록 수정, 응답 쿠키에 토큰 설정 메소드 이름 변경

jaewoo9797 committed:
Refactor: ReservationController 예약 생성하는 로직의 if 분기처리를 Optional과 스트림으로 변경

jaewoo9797 committed:
Refactor: Reservation Service method 추상화

leegwichan commented on review_1_extension.md:
```diff
@@ -0,0 +1,162 @@
+> 서버 측에서 관리하는 비밀키가 유출되지 않는 이상 복호화할 수 없다. 따라서 Signature는 토큰의 위변조 여부를
+> 확인하는데 사용합니다.
+
+클레임은 비밀키를 가지고 있는 서버에서 볼 수 있습니다. 하지만 비밀키가 노출된다면 그 누구라도 복호화하여 인코딩하여 정보를 볼 수 있게 됩니다.
```
![image](https://github.com/user-attachments/assets/ff7832ac-28a4-4c47-bc82-b8901289784d)
![image](https://github.com/user-attachments/assets/52ae1706-5520-45b9-ae79-b2baca674836)
자 이상한 점이 무엇일까요?

leegwichan replied:
> JWT secret key가 외부에 노출 될 경우, 보안적으로 어떤 문제가 있는가?

이 부분에 대해서도 잘 정리해서 보여주세요!

jaewoo9797 replied:
### JWT
- secret Key 가 없어도 복호화 문제 없음
- 그럼 왜 JWT 토큰을 사용하는가?
- secret key 가 노출되면 안되는이유 

직접 jwt 사이트를 방문해서 시도, 비밀키가 없더라도 토큰 안의 값을 읽을 수 있었음.
사용하는 이유는 서버에서 로그인한 클라이언트를 식별하기 위함. 
누구나 볼 수 있고 변조할 수 있기에 secret key를 이용해 변조되었는지를 확인한다. 그렇기에 변조할 수 없도록 노출되어서는 안된다.

해당 내용은 JWT.md 파일에 정리했습니다.

leegwichan commented on review_1_extension.md:
```diff
@@ -0,0 +1,162 @@
+
+그렇다면 특정 엔티티에 속하지 않는 jwt나 결제 로직이라던지 하는 부분들은 무엇인가?
+
+비즈니스 로직인 "유저가 로그인하면 JWT 토큰을 발급해야 한다." 라는 규칙이 존재하는데, jwt 는 단지 토큰을 생성
+, 파싱 해주는 역할을 한다. 이는 도메인 모델에 영향을 주지않는다. 멤버의 데이터에 영향을 주지않는다.
+(데이터베이스에 토큰을 저장하지 않는다고 가정.)
```
DB에 토큰을 저장하는 경우도 있어요. 그럴때는 도메인이라고 할 수 있나요?
e.x. [JWT BlackList](https://velog.io/@junrain2/JWT-BlackList)

jaewoo9797 replied:
DB 에 토큰을 저장하는 것은 회원 도메인를 처리하기 위한 것이 아니라 인증과 권한 관리를 위한 도구일 뿐이기 때문에 도메인이라고 할 수 없지 않나 생각이 들었습니다. 블랙리스트도 결국에는 유저 데이터의 비즈니스 로직을 위한 것이 아니라, 로그아웃이나 토큰 탈취 시에 대응하는 기술적인 도구라고 생각됩니다.

leegwichan commented on review_1_extension.md:
````diff
@@ -0,0 +1,162 @@
+}
+```
+
+### JWT 토큰에서 세션으로 바꾸어야할 떄
+앞 서 UserService에서 AuthService를 SessionService로 교환하게 되면 수정이 조금은 적어질 것 같다. 그러기 위해서는 
+AuthService가 Token 과 Session 방식에서 공통적인 부분을 추상화하여 명세를 잘 뽑아야할 것 같은데, 쉽지는 않은 부분인 것 
+같다.
````
어떻게 구현해야 될 지 보다는 "각 계층 별 책임"에 집중해서 생각해 볼까요?

leegwichan commented on src/main/java/roomescape/config/LoginMemberArgumentResolver.java:
전반적으로 테스트가 없으니까 정상적으로 작동하는지 알기가 힘드네요. 주요 기능 위주로 Rest Assured 테스트 작성 부탁드립니다.

leegwichan commented on src/main/java/roomescape/member/JwtProvider.java:
```diff
@@ -5,22 +5,25 @@
                 .signWith(Keys.hmacShaKeyFor(secretKey.getBytes()))
+                .setHeaderParam(HEADER_TYPE, TOKEN_TYPE)
+                .setSubject(memberResponse.id().toString())
                 .compact();
```
이전에는 name, email을 넣었었는데, 지금은 id값만 넣는 이유는 무엇인가요?

jaewoo9797 replied:
jwt 토큰 안에 어떤 값을 담아주어야 하는지 판단을 못하고 있었습니다. 
수정한 코드에서는,  헤더에 담아온 쿠키 안의 토큰을 이용해 클라이언트 정보를 LoginMember 객체로 변환할 수 있는 값들을 넣어주었습니다.

leegwichan commented on src/main/java/roomescape/member/LoginMember.java:
```diff
@@ -0,0 +1,33 @@
+        this.name = name;
+        this.email = email;
+        this.role = role;
+    }
```
해당 클래스는 record가 아닌 이유는 무엇인가요?

jaewoo9797 replied:
이유는 없었습니다. record 도 하나의 클래스 파일이고 불변형 데이터에 get, equals,hashcode, toString 메소드를 컴파일 시에 생성해주는 것이라고 알고 있습니다.

record 로 바꾸었습니다!

leegwichan commented on src/main/java/roomescape/member/MemberController.java:
```diff
@@ -1,63 +1,63 @@
 
-    public static final String COOKIE_NAME = "token";
+    public static final int ONE_HOUR = 3600;
+    public static final int ZERO = 0;
```
`ZERO` 라고 변수 분리하는 것은 어떤 의미가 있을까요? (단순히 값을 상수로 빼라고 해서 의무적으로 한 건 아니겠죠?)

jaewoo9797 replied:
상수로 관리해야겠다고 생각했었는데, 굳이 뺄 필요는 없는 것 같네요. 상수로 빼서 관리해도 문제는 없을 것 같은데 굳이 뺄 이유도 없네요.

leegwichan commented on src/main/java/roomescape/member/MemberController.java:
```diff
@@ -1,63 +1,63 @@
+        cookieManager.addTokenToCookie(null, ZERO, response);
         return ResponseEntity.ok().build();
     }
+
```
null을 인자로 주고 받는 게 괜찮을까요?

jaewoo9797 replied:
수정했습니다. null로 주고받는 게 괜찮지 않습니다 ㅎㅎ...

leegwichan commented on src/main/java/roomescape/member/MemberDao.java:
```diff
@@ -39,33 +40,29 @@ public Member save(Member member) {
 
     public Optional<Member> findById(Long id) {
-        String query = "SELECT id, name, email, role FROM member WHERE id = ?";
-        return jdbcTemplate.query(query, mapper, id).stream().findFirst();
+        try {
+            String query = "SELECT id, name, email, role FROM member WHERE id = ?";
+            return Optional.ofNullable(jdbcTemplate.queryForObject(query, mapper, id));
+        } catch (EmptyResultDataAccessException exception) {
+            return Optional.empty();
+        }
     }
```
stream 때문에 안 바꾸실 것 같았는데, 다시 queryForObject로 바꾼 이유는 무엇인가요?

jaewoo9797 replied:
저번 리뷰에서 답변 중 리스트 스트림을 이용하는 리소스에 대해서 생각해보았습니다. 
직접 성능 테스트를 할 역량이 되지 않아서 정확한 측정은 못하였습니다. 많은 차이가 없다면 stream 을 쓸 것 같아요 ㅎㅎ...

leegwichan replied:
> 직접 성능 테스트를 할 역량이 되지 않아서 정확한 측정은 못하였습니다. 많은 차이가 없다면 stream 을 쓸 것 같아요 ㅎㅎ...

그러면 안바꾸셔도 됩니다~ 누누히 말씀드리지만, 정답을 찾는 것이 아니라 자신의 의견을 근거를 가지고 주장하는 것이 중요합니다.

leegwichan commented on src/main/java/roomescape/reservation/ReservationService.java:
```diff
@@ -1,21 +1,56 @@
+        Reservation reservation = saveReservation(reservationRequest, foundMember);
+
+        return toReservationResponse(reservation);
+    }
```
reservationRequest에 name이 있는 경우에 아래 것을 쓰면 어떻게 되나요? 메서드 시그니처가 혼동을 줄 수 있게 설계된 것 아닐까요?

jaewoo9797 replied:
### 예약 생성 기능 변경
예약 생성 시 ReservationReqeust의 name이 없는 경우 Cookie에 담긴 정보를 활용하도록 리팩터링 합니다.
- ReservationReqeust에 name값이 있으면 name으로 Member를 찾고
- 없으며 로그인 정보를 활용해서 Member를 찾도록 수정합니다.


요구사항에 맞게 구현하려고 했는데, 보다 나은 방식을 생각해봤는데 기막힌 해결방법이 생각이 안나네요. 하나로 메소드 시그니처를 통합한 다음에 null 체크를 통해 분기 처리하는게 시그니처 혼동이 오지 않을까 싶네요.

leegwichan replied:
위와 아래 메서드는 각각 "어드민을 위한 예약 저장 기능", "일반 사용자를 위한 예약 저장 기능"으로 이해했습니다. (아마 맞겠죠?) 그런데 해당 표현이 없으니, 예약을 생성하고 싶을 때 어떤 메서드를 써야할 지 모를 것 같아요. 뭐 여러가지 방안이 있지만 제 생각을 적어볼께요.

1. 아주 단순하게는 saveForAdmin(), saveForUser() 정도로 구분할 수 있겠죠. (당연히 해당 멤버의 권한을 검증하는 코드가 있어야 하구요.)
2. 아니면 메서드를 하나로 통합해도 될 것 같은데요?
```java
    public ReservationResponse save(ReservationRequest reservationRequest, LoginMember loginMember) {
       if (loginMember.isNotAdmin() && !reservationRequest.name().equals(loginMember.name()) {
            throw RuntimeException();
        }

        Member foundMember = getMemberById(loginMember);
        Reservation reservation = saveReservation(reservationRequest, foundMember);
        return toReservationResponse(reservation);
    }
```

leegwichan commented on 인증_인가_정리.md:
````diff
@@ -0,0 +1,104 @@
+해당 Jwt 로직이 있다고 가정하겠습니다. 만약 세션으로 바꾸게 된다면은 AuthService 를 인터페이스로 바꾸고 명세를 작성한 다음
+SessionAuthService 로 구현체를 만들어서 다음과 같이 구현하거나 현재 서비스 클래스 파일을 변경하면 된다.
+```java
+@Service 
+public class AuthService {
+    public UserResponse login(HttpServletRequest request, HttpServletResponse response, String email, String password) {
+        // 도메인 있는지 처리
+
+        // request 에서 세션을 가지고온다.
+        
+        // 세션에 유저 정보를 키값으로 특정 정보를 담아준다. -> 세션이 등록되면 자동으로 set-cookie 에 JSESSIONID 에 담겨 반환됨
+
+        // 찾은 유저 정보를 DTO 로 변환해서 리턴
+    }
+}
````
HttpServletRequest, HttpServletResponse는 요청과 응답을 담당하는 객체인데, 이 객체가 서비스까지 넘어온 것은 서비스의 책임이 너무 커지는 것 아닐까요?

jaewoo9797 replied:
문서 작성 시에는 질문 주신 레이어를 고려하지 않았습니다.  Jwt 가 도메인인가에 대해서 생각을 정리를 하려고 했었습니다. 

생각을 정리하면서 글을 써서 어지러운데 결론은 Jwt 토큰을 생성하고 파싱하는 것은 도메인, 비즈니스 규칙이 아니라고 생각했습니다.

`도메인은 비즈니스 규칙을 포함하여 해당 객체들이 상호작용하고 데이터를 처리하는지를 포함한 개념이다.`
Jwt는 단순히 인증과관한관리를 위한 기술적인 도구라고 생각합니다.

leegwichan replied:
이 부분 관련해서는 리팩토링 한 부분 잘 봤습니다. 충분히 납득가능하게 분리하신 것 같아요. 👍 
추가적으로 코멘트를 남길까 했지만, 취향의 영역이기도 하고 다른 코멘트에 집중하시는 게 좋을 것 같아 남기지는 않았습니다.

leegwichan reviewed (changes requested):
안녕하세요 곰곰! 커찬입니다.
전반적인 기능 구현 자체는 잘 된 것 같아요. 그런데 테스트 코드가 없어서 잘 작동하는지 확인하기 힘드네요. (시간이 없어서 어쩔 수 없을 것 같지만;;)
"이 로직은 어디에 있어야 할까?"에 대한 고민이나 JWT의 이해도 위주로 코멘트 남겼습니다. 리뷰 반영 & 코멘트 부탁드릴께요!
c.f. 혼자서 고민하시면 어려울거라 생각해요. 자신만의 일차적인 결론을 내보고 다른 사람과 공유하면서 이야기를 나눠보는 것도 좋을 것 같습니다. (저한테 DM 주셔도 되구요)

jaewoo9797 committed:
Docs: JWT 관련 정리

발급된 jwt 토큰은 secretKey 없이도 내부 값을 확인할 수 있는점, 왜 사용하는지, secret key 사용이유에 대해서 이해한 내용을 정리

jaewoo9797 committed:
Refactor: JwtProvider 예외 상황에 따른 처리, 토큰 claim 구성 변경

jaewoo9797 committed:
Feature: JWT 인증 예외 관련 메소드 ExceptionController에 추가

jaewoo9797 committed:
Refactor: LoginMember 클래스를 record 로 변경하고 role 이 Admin 인지 검증 메소드 추가

jaewoo9797 committed:
Refactor: LoginMemberArgumentResolver 디비 접속하는 로직을 삭제하고, 토큰에서 로그인한 유저 정보를 추출하도록 변경

jaewoo9797 committed:
Refactor: Interceptor 에서 로그인한 유저의 정보를 추출한 후 role 이 ADMIN 인지 검사하도록 수정

jaewoo9797 committed:
Refactor: Member Controller 에서 로그인 체크를 하는 엔드 포인트 jwt 토큰에서 추출한 로그인한 유저 정보를 반환하도록 변경

jaewoo9797 committed:
Refactor: Cookie Manager 쿠기 생성 로직 변경

jaewoo9797 committed:
Refactor: AuthService 로그인 한 정보를 추출하는 로직에서 repository 접속하는 로직 삭제

token 의 유저 정보를 이용해서 응답 생성하도록 변경했습니다.

jaewoo9797 committed:
Refactor: MemberResponse 생성 시에 Role 포함하도록 변경

jaewoo9797 committed:
Refactor: 코드 포맷터를 이용한 줄 바꿈

jaewoo9797 committed:
Test: Jwt Provider 예외 스펙변경에 따른 테스트 코드 수정

jaewoo9797 commented on src/main/java/roomescape/member/LoginMember.java:
```diff
@@ -0,0 +1,8 @@
+package roomescape.member;
+
+public record LoginMember(Long id, String name, String email, String role) {
+
+    public boolean isAdmin() {
+        return this.role.equals("ADMIN");
+    }
+}
```
https://github.com/next-step/spring-basic-roomescape-playground/pull/134#discussion_r1994861258

이유는 없었습니다. record 도 하나의 클래스 파일이고 불변형 데이터에 get, equals,hashcode, toString 메소드를 컴파일 시에 생성해주는 것이라고 알고 있습니다. 

record 로 바꾸었습니다!

leegwichan replied:
나중에 기회가 된다면 role을 Enum으로 바꿔보는 것도 좋을 것 같네요.

jaewoo9797 commented on src/main/java/roomescape/member/JwtProvider.java:
```diff
@@ -1,36 +1,73 @@
 
-    public String createToken(MemberResponse memberResponse) {
+    public String generateToken(MemberResponse memberResponse) {
+        Map<String, String> claims = new HashMap<>();
+        claims.put("name", memberResponse.name());
+        claims.put("email", memberResponse.email());
+        claims.put("role", memberResponse.role());
```
https://github.com/next-step/spring-basic-roomescape-playground/pull/134#discussion_r1994859451

jwt 토큰 안에 어떤 값을 담아주어야 하는지 잘 몰랐습니다. 저는 이번에 담아준 값들은 헤더에 담아온 쿠키 안의 토큰을 까서 클라이언트 정보를 LoginMember 객체로 변환할 수 있는 값들을 넣어주었습니다.

leegwichan replied:
일단 JWT는 충분히 다른 사람에게 탈취당할 수 있는 정보이죠? 그러면 쿠키를 탈취당한다면 해당 사람의 "이름", "이메일", "서비스 내 권한"이 노출되겠네요. 곰곰은 특정 서비스를 사용하는데 임의의 사용자에게 해당 정보가 노출된다면, 해당 서비스를 쓰기 꺼려지게 되겠죠?

일단 JWT에 어떤 정보를 담는다는 절대적인 기준은 없습니다. 하지만 유저 정보는 최소한으로 담아야 유저의 개인 정보 노출을 줄일 수 있을 것 같아요. 예를 들면, 해당 유저를 알 수 있는 고유 정보(id, email)을 하나 담고, LoginMember를 만들 때 DB에 member를 조회해서 나머지 정보를 가져오는 방법이 있겠죠.

jaewoo9797 commented on src/main/java/roomescape/member/MemberController.java:
```diff
@@ -19,9 +19,6 @@
 public class MemberController {
 
-    public static final int ONE_HOUR = 3600;
-    public static final int ZERO = 0;
```
상수로 관리해야겠다고 생각했었는데, 굳이 뺄 필요는 없는 것 같네요. 상수로 빼서 관리해도 문제는 없을 것 같은데 굳이 뺄 이유도 없네요

jaewoo9797 reviewed (commented):
커찬 안녕하세요.
좋지 않은 사정으로 인해서 현재 작업한 부분까지 댓글 남겨두었습니다.

jaewoo9797 committed:
Refactor: ReservationController주석제거, 예외 클래스 수정 ReservationService

jaewoo9797 committed:
Test : Rest Assured 사용해서 엔드 포인트 테스트

jaewoo9797 committed:
Test : Rest Assured 사용해서 엔드 포인트 테스트

leegwichan commented on src/main/java/roomescape/member/MemberService.java:
```diff
@@ -1,59 +1,32 @@
-    }
-
-    private MemberResponse findByEmailAndPassword(String email, String password) {
-        Member member = getMemberByEmailAndPassword(email, password);
-        return toMemberResponse(member);
+                new Member(memberRequest.name(), memberRequest.email(), memberRequest.password(), DEFAULT_ROLE));
```
`DEFAULT_ROLE = "USER"`라는 사실은 도메인이 알고 있는 편이 더 좋을 것 같아요. role을 인자로 받지 않는 생성자를 하나 더 만들어서 사용하는 편도 좋겠네요.

leegwichan commented on src/main/java/roomescape/member/enums/Role.java:
```diff
@@ -0,0 +1,18 @@
+package roomescape.member.enums;
+
+public enum Role {
+
+    ADMIN("관리자"),
+    USER("사용자")
+    ;
+
+    private final String description;
```
description 은 사용하지 않는다면 굳이 필요 없을 것 같아요.

leegwichan commented on src/main/java/roomescape/reservation/ReservationService.java:
```diff
@@ -1,21 +1,57 @@
+    }
 
-        return new ReservationResponse(reservation.getId(), reservationRequest.getName(), reservation.getTheme().getName(), reservation.getDate(), reservation.getTime().getValue());
+    public ReservationResponse save(ReservationRequest reservationRequest, LoginMember loginMember) {
+        Member foundMember = getMemberById(loginMember);
+
+        Reservation reservation = saveReservation(reservationRequest, foundMember);
+
+        return toReservationResponse(reservation);
+    }
```
각 명령어 사이에 빈 공백은 의미있게 사용해주세요. 매 명령어마다 공백을 만들어 쓰면 가독성만 떨어진다고 생각합니다.

leegwichan commented on src/test/java/roomescape/reservation/ReservationControllerTest.java:
테스트 잘 작성하셨네요 👍 
조금 더 공부하실 것이라면 [테스트 격리](https://hudi.blog/isolated-test/)에 대해 더 알아보면 좋을 것 같아요.

leegwichan reviewed (approved):
안녕하세요 곰곰! 커찬입니다.
예상치 못한 상황 때문에 리뷰 반영하는데 긴 시간이 걸렸네요. JWT 관련해서 여러 관점으로 생각해보고 리팩토링한 것이 인상 깊었습니다. 그래서 그런지 큰 문제가 되는 부분은 언뜻 봐서는 없었습니다.
이전 코멘트에서 반영한 내용과 일부 눈에 띄는 몇 군데 위주로 리뷰 남겼습니다. 천천히 살펴보시고 참고하시면 좋을 것 같아요. 다음 미션은 DB 접근 기술을 다루기 때문에 제가 리뷰한 모든 내용을 반영하실 필요는 없을 것 같아요. 다음 미션에 지장이 될 부분만 반영하시면 되겠습니다.

바쁘신 와중에 고생하셨습니다. 이만 머지하도록 할께요. 다음 미션도 화이팅!
